lib/src/main/scala/org/typelevel/scalacoptions/proposal/ScalacOption.scala

```diff
@@ -0,0 +1,118 @@
+ * limitations under the License.
+ */
+
+package org.typelevel.scalacoptions.proposal
+
+import org.typelevel.scalacoptions.ScalaVersion
+
+trait ScalacOption[A] { // }, T <: ScalacOption.Type] {
+  type Type <: ScalacOption.Type
+
+  // ..or `isSupportedFor` maybe?
+  def isSupported(sv: ScalaVersion): Boolean
+
+  /** Base option name.
+    *
+    * @return
+    *   - for `"-deprecation"` results to `"deprecation"`
+    *   - for `"-deprecation:true"` results to `"deprecation"`
+    *   - for `"-Xlint"` results to `"Xlint"`
+    *   - for `"-Xlint:deprecation"` results to `"Xlint"`
+    */
+  def baseName: String
+
+  // /** Full option name.
+  //   *
+  //   * @return
+  //   *   - for `"-deprecation"` results to `"deprecation"`
+  //   *   - for `"-deprecation:true"` results to `"deprecation"` (same as above!)
+  //   *   - for `"-Xlint"` results to `"Xlint"`
+  //   *   - for `"-Xlint:deprecation"` results to `"Xlint:deprecation"`
+  //   */
+  // def fullName: String
+
+  /** Parses the option from a raw value found for the option's base name.
+    *
+    * @param rawValue
+    *   a raw option value, can be empty strings for options like `-deprecation`
+    *
+    * @return
+    *   - `None` if the value does not belong to the option and thus cannot be parsed, i.e. when
+    *     `"unused"` is passed where `-Xlint:deprecation` model is expected;
+    *   - `Some(A)` when a correct value is passed for the option, e.g. both `"unused"` and
+    *     `"-unused"` will be accepted for the an option type class that models the `-Xlint:unused`
+    *     option.
+    *
+    * @note
+    *   It deliberately made not returning any failure type like `ParseFailure`, becase
+    *   `ScalacOption.Select` is responsible for that.
+    */
+  def parse(rawValue: String): Option[A]
+}
+
+object ScalacOption {
+  type Aux[A, T <: Type] = ScalacOption[A] { type Type = T }
+
+  // Note: initially I thought that Singe/Recurring might not be necessary.
+  //       But later I realized that `Recurring` can be useful for modeling such options like "-Wconf"
+  sealed trait Type
+  abstract final class Single private ()    extends Type // never instantiated
+  abstract final class Recurring private () extends Type // never instantiated
+
+  sealed trait Select[A] {
+    // Not sure why it is a higher kinded type in http4s' Header.
+    // A regular plain type seems to be just enough here.
+    type Out
+
+    /** Parses all raw option values groupped by their [[ScalacOption.baseName]].
+      *
+      * @param rawValues
+      *   a list of raw option values in the order they occured in a command line.
+      */
+    def parse(rawValues: List[String]): Option[Out]
+  }
+
+  object Select {
+    implicit def singleScalacOption[A](implicit
+      so: ScalacOption.Aux[A, Single]
+    ): Select[A] { type Out = A } =
+      new Select[A] {
+        type Out = A
+
+        def parse(rawValues: List[String]): Option[A] = {
```

Comment:
I imagine this method will need to delimit the scalac options without knowing all of them. One easy algorithm is to treat every `-something` as the beginning of a new option+args.

Reply:
I would say that a more reliable (and performant, actually) way could be pre-parsing all the options at once, and then just picking them by one from the collection of parsed options upon a user request. But it would require a global "registry" of parsers. Which is not that bad, actually, but makes the approach different from how Headers are treated.

The root cause of the difference is kinda "fundamental" – in regards to headers there're RFCs that clearly define how to separate various headers from each other without digging into their contents. But for CLI options there's no such a clear way of separating them 🤷

Reply:
> But it would require a global "registry" of parsers.

I agree it would be much more reliable, but I would very much like to avoid this for fear this library will completely stop working if it encounters an unknown scalac option. It must be able to cleanly pass-through anything it cannot recognize.

Reply:
Right, we definitely should try to make the parser more lenient.
But that should be a fall-back mode, I think. I.e. if the parser cannot recognize something it can try to pass-through assuming that everything not starting with a dash is an argument rather than an option.

It would be a good trade-off between reliability and tolerance.